test/e2e/vertical_scaling_test.go

```diff
@@ -111,44 +119,112 @@ func ensureMasterMachineRunning(ctx context.Context, t testing.TB, machineName s
+func ensureRunningMachinesAndCount(ctx context.Context, t testing.TB, machineClient machinev1beta1client.MachineInterface) {
+	err := checkRunningMachinesAndCount(ctx, machineClient)
+	require.NoError(t, err)
+}
+
+// checkRunningMachinesAndCount checks if there are only 3 running master machines otherwise it returns an error
+func checkRunningMachinesAndCount(ctx context.Context, machineClient machinev1beta1client.MachineInterface) error {
+	machineList, err := machineClient.List(ctx, metav1.ListOptions{LabelSelector: masterMachineLabelSelector})
+	if err != nil {
+		return err
+	}
 
 	if len(machineList.Items) != 3 {
-		machineNames := []string{}
+		var machineNames []string
 		for _, machine := range machineList.Items {
 			machineNames = append(machineNames, machine.Name)
 		}
-		t.Fatalf("incorrect cluster size, expected exactly 3 master machines, got: %v, found: %v", len(machineList.Items), machineNames)
+		return fmt.Errorf("expected exactly 3 master machines, got %d, machines are: %v", len(machineList.Items), machineNames)
 	}
 
 	for _, machine := range machineList.Items {
 		machinePhase := pointer.StringDeref(machine.Status.Phase, "")
 		if machinePhase != "Running" {
-			t.Fatalf("%q machine is in unexpected %q state, expected Running", machine.Name, machinePhase)
+			return fmt.Errorf("%q machine is in unexpected %q state, expected Running", machine.Name, machinePhase)
 		}
 	}
+	return nil
 }
 
-// ensureMembersCount simply counts the current etcd members, it doesn't evaluate health conditions or any other attributes (i.e. name) of individual members
+func recoverClusterToInitialStateIfNeeded(ctx context.Context, t testing.TB, machineClient machinev1beta1client.MachineInterface) error {
+	machineList, err := machineClient.List(ctx, metav1.ListOptions{LabelSelector: masterMachineLabelSelector})
+	if err != nil {
+		return err
+	}
+
+	var machineNames []string
+	for _, machine := range machineList.Items {
+		machineNames = append(machineNames, machine.Name)
+	}
+
+	t.Logf("checking if there are any excessive machines in the cluster (created by a previous test), expected cluster size is 3, found %v machines: %v", len(machineList.Items), machineNames)
+	for _, machine := range machineList.Items {
+		if strings.HasSuffix(machine.Name, "-clone") {
+			err := machineClient.Delete(ctx, machine.Name, metav1.DeleteOptions{})
+			if err != nil {
+				return fmt.Errorf("failed removing the machine: %q, err: %v", machine.Name, err)
+			}
+			t.Logf("successfully deleted an excessive machine %q from the API (perhaps, created by a previous test)", machine.Name)
+		}
+	}
+
+	return nil
+}
+
+// ensureMembersCount same as checkMembersCount but will fail on error
 func ensureMembersCount(t testing.TB, etcdClientFactory etcdClientCreator, expectedMembersCount int) {
+	err := withRetries(t, 2, func() error { return checkMembersCount(t, etcdClientFactory, expectedMembersCount) })
```

Comment:
Can you please expand on why we're retrying `checkMembersCount()` twice? And each retry seems to be spaced 40s apart?
Are we expecting transient errors?

`checkMembersCount()` already has polling built-in for 5mins. And it seems to be already retrying on transient errors (e.g client connect fail). So this just seems like another knob to double the overall polling duration.

If that's the case then is it possible if we could just handle the retry and polling configuration for `ensureMembersCount()` within `checkMembersCount()` perhaps as a parameter. To me that makes the retry behavior easier to reason about rather than having 2 places to control the effective retry and polling duration.

Reply:
> Are we expecting transient errors?

yes, we are, atm removing a member is a destructive operation and it is hard to say how long it will take to recover the cluster.

sure, if you don't like using `withRetries` I could extend the polling timeout in `checkMembersCount` method.

Reply:
I primarily introduced `withRetries` to wrap `ensureInitialClusterState`. This method cleans up the cluster state left in an unexpected state by the previous test(s).